asteroid/engine/gan_system.py

```diff
@@ -0,0 +1,161 @@
+        self.generator = generator
+        self.opt_d = opt_d
+        self.opt_g = opt_g
+        self.d_loss = discriminator_loss
+        self.g_loss = generator_loss
+        self.validation_loss = validation_loss
+        self.train_loader = train_loader
+        self.val_loader = val_loader
+        self.scheduler_d = scheduler_d
+        self.scheduler_g = scheduler_g
+        self.conf = conf
+
+    def forward(self, z):
+        return self.generator(z)
+
+    def routine_step(self, batch, optimizer_idx):
+
+        # Get data from data_loader
+        inputs, targets = batch
+        # Forward inputs
+        estimates = self(inputs)
+        # The discriminator is basically a binary classifier
+        true_labels = True
+        false_labels = False
+
+        # Train discriminator
+        if optimizer_idx == 0:
+            # Compute D loss for targets (labels = 1)
+            est_true_labels = self.discriminator(targets, inputs, targets)
+            true_loss = self.d_loss(inputs, targets, estimates,
+                                    est_true_labels, true_labels)
+            # Compute D loss for estimates (labels = 0)
+            est_false_labels = self.discriminator(estimates.detach(), inputs,
+                                                  targets)
+            fake_loss = self.d_loss(inputs, targets, estimates,
+                                    est_false_labels, false_labels)
+            # Overall, the loss is the mean of these
+            d_loss = (true_loss + fake_loss) * 0.5
+            tqdm_dict = {'d_loss': d_loss}
+            output = OrderedDict({
+                'loss': d_loss,
+                'progress_bar': tqdm_dict,
+                'log': tqdm_dict
+            })
+            return output
+
+        # Train generator
+        if optimizer_idx == 1:
+            # The generator is supposed to fool the discriminator.
+            # To do so labels are set as true.
+            est_labels = self.discriminator(estimates, inputs, targets)
+            adversarial_loss = self.g_loss(estimates, targets, est_labels)
+            tqdm_dict = {'g_loss': adversarial_loss}
+            output = OrderedDict({
+                'loss': adversarial_loss,
+                'progress_bar': tqdm_dict,
+                'log': tqdm_dict
+            })
+            return output
+
+    def training_step(self, batch, batch_nb, optimizer_idx):
+        output = self.routine_step(batch, optimizer_idx)
+        return output
+
+    def validation_step(self, batch, batch_nb):
+        """ Need to overwrite PL validation_step to do validation.
+
+        Args:
+            batch: the object returned by the loader (a list of torch.Tensor
+                in most cases) but can be something else.
+            batch_nb (int): The number of the batch in the epoch.
+
+        Returns:
+            dict:
+
+            ``'val_loss'``: loss
+        """
+        inputs, targets = batch
```

Comment:
By default, don't we want d_loss and g_loss at validation time? 
I'm no GAN expert, just asking.

Reply:
In the validation step we only pass data through the generator and evaluate its performance.
The value of d_loss and g_loss is not very relevant at this point because

Reply:
Because what?

Reply:
Ahah sry. Because they don't give information about the performance, and they will converge to a value after few iterations, so I don't think it's worth the computation